---
params:
  projectdir: !r here::here() # Can hard code if for some reason you need that
  project_name: "Default"     # Change depending on your project name
  species: "rat"                # one of human, mouse, rat, hamster
  design: "Group"             # single experimental group of interest; entries in this column must match the contrast names.
  intgroup: ["Group"]           # experimental group of interest plus covariates; can be more than one
  nuisance:  !r NA #"Batch" To remove batch effects for PCA/Heatmaps visualiziations
  flag: TRUE                  # runs all analysis by default when set to TRUE
  platform: "RNA-Seq"         # RNA-Seq or TempO-Seq
  group_facet: "Timepoint" #!r NA          # If you have many different experimental groups, you may subset the report by specifying a column in the metadata to filter groups, and then setting the group of interest in group_filter
  group_filter:  "5D" # !r NA # Which group will this report be done on?
  strict_contrasts: TRUE # Use both columns (exp, cont) of contrasts file to limit what is included in the report
  exclude_samples: !r NA  # Optionally, a vector of sample names to exclude from the analysis
  exclude_groups: !r NA # Optionally, a vector of groups to exclude from the analysis. By default this is assumed to be in the column specified by params$design.
  include_only_column:  !r NA # Restrict analysis to group(s) in the column listed here based on params$include_only_group.
  include_only_group:  !r NA # Restrict analysis to this/these group(s) within the column listed in params$include_only_column
  use_cached_RData: FALSE # If possible, load the saved RData for dds object and gene IDs
  cpus: 39 # Set to a lower number (e.g., 2 to 4) if you aren't working in a server environment
  run_pathway_analysis: TRUE # Optionally disable pathway analysis if not available for your organism
  wikipathways_directory: "~/shared/dbs/wikipathways"
  biospyder_dbs: "~/shared/dbs/biospyder/"
  nBestFeatures: 20 # The number of best features to make plots of their counts
  nBest: 100 # Number of features to include in table and limiting PCA/clustering analysis
  nHeatmap: 50 # Number of most variable genes for heatmap
  nHeatmapDEGs: 50 # Number of DEGs for heatmap
output:
  html_document:
    toc: true
    toc_float: true
    number_sections: true
    code_folding: hide
    theme: spacelab           # flatly spacelab sandstone cerulean
    code_download: true
bibliography: "`r file.path(params$projectdir, 'Rmd/references.bib')`"
title: "`r paste('Transcriptome analysis:', gsub(pattern = '_', replacement = ' ', x = params$project_name))`"
subtitle: "`r paste(params$platform, 'analysis', ifelse(!is.null(params$group_filter), paste('for', paste(params$group_filter, collapse=' and ')), ''))`"
author: "Your Name"
---

***

# Revision History

**This version**  
  
*Current version: 1*  
Date report generated: `r format(Sys.time(), '%d %B, %Y')`  
Report prepared for: Name  
Purpose of report:  

* Exercise to analyze RNA-Seq data  

**Previous revisions**  

N/A  

***

# Code setup, Data Loading, and DESeq2 Code

This report is meant to help explore DESeq2 results and was generated using RMarkdown. This section contains the code for setting up the rest of the report.  

## Load libraries

```{r docSetup, warning = FALSE, message = FALSE}
## Define paths
paths <- list()
# Other important system paths to specify in config
paths$wikipathways <- params$wikipathways_directory
# For project structure
paths$root <- params$projectdir
paths$data <- paste0(paths$root, "/data")
paths$raw <- paste0(paths$data, "/raw")
paths$output <- paste0(paths$data, "/output")
paths$processed <- paste0(paths$data, "/processed")
paths$metadata <- paste0(paths$root, "/metadata")
paths$reports <- paste0(paths$root, "/reports")
if (is.na(params$group_facet)) {
  paths$DEG_output <- paste0(paths$root, "/DEG_output")
  paths$pathway_analysis <- paste0(paths$DEG_output, "/pathway_analysis")
} else {
  paths$DEG_output <- paste0(paths$root,
                             "/DEG_output/group_",
                             paste(params$group_filter, collapse = "_"))
  paths$pathway_analysis <- paste0(paths$DEG_output, "/pathway_analysis")
}
paths$RData <- paste0(paths$DEG_output, "/RData")

## knitrBoostrap and device chunk options
# Set so that long lines in R will be wrapped:
knitr::opts_chunk$set(bootstrap.show.code = FALSE,
                      bootstrap.panel = TRUE,
                      cache = FALSE)
#knitr::tidy.opts = list(width.cutoff = 60), tidy = TRUE, crop = NULL # Use for PDFs
knitr::opts_knit$set(root.dir = paths$root)

# Set this variable to FALSE if you only want to run the plotting functions by reloading the last RData file.
flag = params$flag # To evaluate analysis chunks
if (flag == FALSE) { load(paste0(params$projectdir, "Partial_analysis.RData")) }
# Set this variable to TRUE if you would like to embed the files directly into the HTML for portability. This slows down page responsiveness drastically, since the files are generally quite large.
embedFiles = FALSE
# Se this variable to be TRUE if you want to have separate plots of top genes as defined in the R-ODAF template
R_ODAF_plots = FALSE
pathway_analysis <- file.path(paths$root, "Rmd/Pathway_analysis.Rmd")
```

```{r 'warn_user', eval=!flag}
message("If this text is visible by default, this report was produced to test plotting functions and should be used exclusively for testing and development.")
```

```{r setup, warning = FALSE, message = FALSE}
#### Record start time
startTime <- Sys.time()

#### Load Libraries
# General purpose
# library('conflicted')
library('knitr')
library('kableExtra')
library('tidyverse')
library('magrittr')
library('ggplot2')
library('DT')
library('data.table')
library('pheatmap')
library('lattice')
library('tidytext')
library('openxlsx')
library('RColorBrewer')
library('viridis')
library('sessioninfo')
library('plotly')
library('dendextend')
#library('RMariaDB')

# Bioinformatics-specific
library('DESeq2')
library('limma')
library('edgeR')
library('enrichplot')
library('rWikiPathways')
library('BiocParallel')
library('clusterProfiler')
library('biomaRt')
library('AnnotationHub')
library('vsn')

# Load genome, depends on species
# To add: Zebrafish, Folsomia candida, others?
species <- params$species
if (species == "human") {
  # Human:
  library('org.Hs.eg.db')
  orgdb <- "org.Hs.eg.db"
  species_sci <- "Homo sapiens"
  wiki <- "wikipathways-20210610-gmt-Homo_sapiens.gmt"
  ensembl_species <- "hsapiens_gene_ensembl"
  species_gene_symbol <- "external_gene_name"
  kegg_organism <- "hsa"
  temposeq_manifest <- "191113_Human_S1500_Surrogate_2.0_Manifest.csv" # options: "191113_Human_S1500_Surrogate_2.0_Manifest.csv" "191004_Human_Whole_Transcriptome_2.0_Manifest.txt"
} else {
  if (species == "mouse") {
    # Mouse:
    library('org.Mm.eg.db')
    orgdb <- "org.Mm.eg.db"
    species_sci <- "Mus musculus"
    wiki <- "wikipathways-20210610-gmt-Mus_musculus.gmt"
    ensembl_species <- "mmusculus_gene_ensembl"
    species_gene_symbol <- "mgi_symbol"
    kegg_organism <- "mmu"
    temposeq_manifest <- "181130 Mouse S1500+ Surrogate 1.2 Manifest.xlsx" # or 190603 Mouse Whole Transcriptome 1.1 Manifest.xlsx
} else {
  if (species == "rat") {
    # Rat: 
    library('org.Rn.eg.db')
    orgdb <- "org.Rn.eg.db"
    species_sci <- "Rattus norvegicus"
    wiki <- "wikipathways-20210610-gmt-Rattus_norvegicus.gmt"
    ensembl_species <- "rnorvegicus_gene_ensembl"
    species_gene_symbol <- "rgd_symbol"
    kegg_organism <- "rno"
    temposeq_manifest <- "190809 Rat Whole Transcriptome 1.0 Manifest.xlsx"
} else {
  if (species == "hamster") {
    # Golden hamster:
    OrgDb.Ma <- query(AnnotationHub(), c("OrgDb", "Mesocricetus auratus"))[[1]]
    orgdb <- "OrgDb.Ma"
    species_sci <- "Mesocricetus auratus"
    ensembl_species <- "mauratus_gene_ensembl"
    species_gene_symbol <- "external_gene_name"
} else {
  message("No species picked!")
}
}
}
}

# Options
options(java.parameters = "-Xmx10000m")
theme_set(theme_bw())
```

The code above (not shown by default) loads the relevant R packages required for analysis.

## Specify parameters

```{r setup_code, warning = FALSE, message = FALSE, collapse = TRUE}
################################################################################
################################################################################
# PARAMETERS - should be set by YAML header.
################################################################################
################################################################################

# Various plotting and display options
nBestFeatures <- params$nBestFeatures
nBest <- params$nBest
nHeatmap <- params$nHeatmap
nHeatmapDEGs <- 50

# Set analysis ID. This ID will be used as prefix for the output files
# Normally, as follows: year - project_name - group_filter
if (is.na(params$group_filter)) {
  analysisID <- paste(format(Sys.time(), '%Y'), params$project_name, sep = "_")
} else {
  analysisID <- paste(format(Sys.time(), '%Y'),
                      params$project_name,
                      paste(params$group_filter, collapse = "_"),
                      sep = "_")
}
# Specify used platform/technology for data generation:
Platform <- params$platform # Should be one of "RNA-Seq" or "TempO-Seq"
NORM_TYPE <- paste0(analysisID, "_DESeq2_", Platform)
# Misc parameters
digits = 2 # For rounding numbers

################################################################################
################################################################################
#### FILES TO LOAD
################################################################################
################################################################################
# A. Tab delimited file with merged RSEM.genes.results files:
if (Platform == "TempO-Seq") {
  SampleDataFile <- file.path(paths$processed, "count_table.csv")
  sampledata_sep = ","
} else {
  SampleDataFile <- file.path(paths$processed, "genes.data.tsv")
  sampledata_sep = "\t"
}
  
# B. Tab delimited sample information file with at least 2 columns:
  # 1. sample names identical to the column names of sampleData
  # 2. compound/group/whatever (needs to identify to which experimental group the sample belongs)
SampleKeyFile <- file.path(paths$metadata, "metadata.QC_applied.txt")
# C. Tab delimited file of comparisons (contrasts) to test 
# Group of interest in the left column, control for comparison in the right column
ContrastsFile <- file.path(paths$metadata, "contrasts.txt")

# Set file locations
if (!dir.exists(paths$DEG_output)) {dir.create(paths$DEG_output, recursive = TRUE)}
if (!dir.exists(paths$pathway_analysis)) {dir.create(paths$pathway_analysis, recursive = TRUE)}
if (!dir.exists(paths$RData)) {dir.create(paths$RData, recursive = TRUE)}

# Specify which groups need to be compared 
contrasts <- read.delim(ContrastsFile, stringsAsFactors = FALSE, sep = "\t", header = FALSE,  quote = "\"")
DESIGN <- params$design	# Column name which defines the groups to be compared
intgroup <- params$intgroup # "Interesting groups" - experimental group/covariates

```

The code above (not shown by default) specifies user preferences and data locations.

## Load data

```{r 'load_data', collapse = TRUE, warning = FALSE, message = FALSE, eval = flag}
# Load input files
sampleData <- read.delim(SampleDataFile,
                         sep = sampledata_sep,
                         stringsAsFactors = FALSE,
                         header = TRUE, 
                         quote = "\"",
                         row.names = 1,
                         check.names = FALSE)

DESeqDesign <- read.delim(SampleKeyFile,
                          stringsAsFactors = FALSE,
                          sep = "\t",
                          header = TRUE,
                          quote = "\"",
                          row.names = 1) # Pick column that is used in ID; might be more appropriate to change this!
DESeqDesign$original_names <- rownames(DESeqDesign)
DESeqDesignAsRead <- DESeqDesign

# Parameter-specified exclusions
# Conditionally exclude contrasts and metadata that should be filtered
# 1. Filter by group filter - useful for selecting one treatment at a time. params$group_filter
# 2. Filter manually - useful for removing irrelevant control groups (e.g. reference RNA). params$exclude_groups
# Note that these should only be run if the respective variables are set.
if (!is.na(params$group_facet)) {
  contrasts_to_filter <- DESeqDesign %>%
  dplyr::filter(!!sym(params$group_facet) %in% params$group_filter) %>% # NOTE: Not sure if %in% or == is better here.
  pull(params$design) %>% 
  unique()
  contrasts <- contrasts %>% dplyr::filter(V1 %in% contrasts_to_filter)
  if (params$strict_contrasts == T) {
    contrasts <- contrasts %>% dplyr::filter(V2 %in% contrasts_to_filter)
  }
  DESeqDesign <- DESeqDesign %>% dplyr::filter(!!sym(params$design) %in% (unlist(contrasts) %>% unique()) )
}
if (any(!is.na(params$exclude_samples))) {
  DESeqDesign <- DESeqDesign %>% dplyr::filter(!original_names %in% params$exclude_samples)
}
if (any(!is.na(params$exclude_groups))) {
  DESeqDesign <- DESeqDesign %>% dplyr::filter(!(!!sym(params$design)) %in% params$exclude_groups)
  contrasts_to_filter <- DESeqDesign %>% 
  dplyr::filter(!(!!sym(params$design)) %in% params$exclude_groups) %>%
  pull(params$design) %>% 
  unique()
  contrasts <- contrasts %>% dplyr::filter(V1 %in% contrasts_to_filter)
  if (params$strict_contrasts == T) {
    contrasts <- contrasts %>% dplyr::filter(V2 %in% contrasts_to_filter)
  }
}
if (!is.na(params$include_only_column) & !is.na(params$include_only_group)) {
  DESeqDesign <- DESeqDesign %>%
    dplyr::filter((!!sym(params$include_only_column)) %in% params$include_only_group)
  limit_contrasts <- DESeqDesign %>%
    pull(!!sym(params$design)) %>%
    unique() %>%
    as.character()
  contrasts <- contrasts %>% dplyr::filter(V1 %in% limit_contrasts)
}

# If there is a dose column, reorder the contrast list by dose
if (any(grepl(x = colnames(DESeqDesign), pattern = "dose", ignore.case = T))) {
  doseCol <- grep(x = colnames(DESeqDesign), pattern = "dose", ignore.case = T)
  ordered_by_dose <- DESeqDesign[order(DESeqDesign[doseCol]),] %>%
    dplyr::select(params$design) %>%
    dplyr::pull()
  contrasts <- contrasts %>%
    dplyr::slice(match(ordered_by_dose, V1)) %>%
    unique()
}

# Once contrasts have been filtered, generate short contrast names for report.
short_contrast_names <- paste(contrasts$V1, "v.", contrasts$V2) # Customize these for your experiment... Must be short enough to fit as Excel tab titles.

# Create directories
plotdir <- paste(paths$DEG_output, "/plots/", sep = "")
if (!dir.exists(plotdir)) {dir.create(plotdir, recursive = TRUE)}
barplot.dir <- paste(plotdir, "/barplot_genes/", sep = "")
if (!dir.exists(barplot.dir)) {dir.create(barplot.dir, recursive = TRUE)}

#Set parameters according to platform
if (Platform == "RNA-Seq") {
  threshold = 1000000 # Number of aligned reads per sample required
  MinCount <- 1
  alpha <- pAdjValue <- 0.05 # Relaxed from 0.01
  linear_fc_filter <- 1.5
  biomart_filter <- "ensembl_gene_id"
} else if (Platform == "TempO-Seq") {
  threshold = 100000 # Number of aligned reads per sample required
  MinCount <- 0.5
  alpha <- pAdjValue <- 0.05 
  linear_fc_filter <- 1.5
  biomart_filter <- "ensembl_gene_id"
  
  biospyder <- read.delim(file.path(params$biospyder_dbs, temposeq_manifest), # Assay manifest...
                          stringsAsFactors = FALSE,
                          sep = "\t",
                          header = TRUE,
                          quote = "\"")
  
  # Fill in the blanks for TempO-Seq Manifest...
  # Set NULL values to NA
  biospyder[ biospyder == "NULL" ] <- NA
  
  # Separate out the coordinates into useful columns
  biospyder <- biospyder %>%
    separate(PROBE_COORDINATE, into = c("chr","pos","strand"), sep = ":" ) %>%
    separate(pos, into = c("start","end"), sep = "-") %>%
    mutate(start = as.numeric(start)) %>%
    mutate(end = as.numeric(end)) %>%
    mutate(strand = ifelse(strand == "-","-1","1"))
  
  # Make a table of the probes where we need a gene ID
  missing <-  biospyder %>%
    dplyr::filter(is.na(ENSEMBL_GENE_ID))
  
  # Get Ensembl Gene IDs using biomaRt based on locus
  to_add_missing <- list()
  try_again <- list()
  for (row in seq_along(row.names(missing))) {
  #for (row in 1:10) {
    if (is.na(missing[row,]["chr"])) {
      try_again <- c(try_again, missing[row,][["GENE_SYMBOL"]])
    } else {
      id <- getBM(filters = c('chromosome_name','start','end','strand'),
                  attributes = c(biomart_filter,
                                 'chromosome_name',
                                 'start_position',
                                 'end_position',
                                 species_gene_symbol,
                                 "strand"),
                  values = list(missing[row,]["chr"],
                                missing[row,]["start"],
                                missing[row,]["end"],
                                missing[row,]["strand"]),
                  mart = ensembl,
                  uniqueRows = T)
      if (nrow(id) > 0) {
        id$PROBE_ID <- missing[row,][["PROBE_ID"]]
        to_add_missing[[row]] <- id
      } else {
        try_again <- c(try_again, missing[row,][["GENE_SYMBOL"]])
      }
    }
  }
  to_add_missing <- rbindlist(to_add_missing)
  to_add_missing$chr <- as.character(to_add_missing$chromosome_name)
  to_add_missing$strand <- as.character(to_add_missing$strand)
  to_add_missing$GENE_SYMBOL <- to_add_missing$external_gene_name
  to_add_missing <- to_add_missing %>%
    dplyr::select(-c(external_gene_name, chromosome_name))
  
  symbol_to_ensembl <- getBM(filters = c('external_gene_name'),
                             attributes = c(biomart_filter,
                                            'chromosome_name',
                                            'start_position',
                                            'end_position',
                                            species_gene_symbol,
                                            "strand"),
                             values =  unlist(try_again),
                             mart = ensembl,
                             uniqueRows = T)

  symbol_to_ensembl$chr <- as.character(symbol_to_ensembl$chromosome_name)
  symbol_to_ensembl$strand <- as.character(symbol_to_ensembl$strand)
  symbol_to_ensembl$GENE_SYMBOL <- symbol_to_ensembl$external_gene_name
  symbol_to_ensembl <- symbol_to_ensembl %>%
    dplyr::select(-c(external_gene_name, chromosome_name))
  
  try_again_2 <- unlist(try_again)[!unlist(try_again) %in% symbol_to_ensembl$GENE_SYMBOL]
  try_again_2 <- missing[missing$GENE_SYMBOL %in% try_again_entrez,]
  
  synonym_to_ensembl <- getBM(filters = c('external_synonym'),
                              attributes = c(biomart_filter,
                                            'chromosome_name',
                                            'start_position',
                                            'end_position',
                                            species_gene_symbol,
                                            "strand",
                                            "external_synonym"),
                              values =  try_again_2,
                              mart = ensembl,
                              uniqueRows = T)
  
  biospyder <- left_join(biospyder,
                    to_add_missing,
                    by = c("PROBE_ID"),
                    suffix = c("", ".y")) %>%
    mutate(ENSEMBL_GENE_ID = dplyr::coalesce(ENSEMBL_GENE_ID, ensembl_gene_id)) %>%
    dplyr::select(-ensembl_gene_id) %>%
    left_join(symbol_to_ensembl, by = c("GENE_SYMBOL" = "GENE_SYMBOL"), suffix = c("", ".y")) %>% 
    mutate(ENSEMBL_GENE_ID = dplyr::coalesce(ENSEMBL_GENE_ID, ensembl_gene_id)) %>%
    dplyr::select(-ensembl_gene_id) %>%
    left_join(synonym_to_ensembl, by = c("GENE_SYMBOL" = "external_synonym"), suffix = c("", ".y")) %>%
    mutate(ENSEMBL_GENE_ID = dplyr::coalesce(ENSEMBL_GENE_ID, ensembl_gene_id)) %>%
    dplyr::select(-ensembl_gene_id)
  biospyder <- biospyder %>% dplyr::select(1:12)
  
  # biospyder %>% dplyr::filter(is.na(ENSEMBL_GENE_ID)) %>% glimpse()
  
  #################################################
  # Don't do this
  # It collapses probes by gene sybmols - which is not really the best way.
  # We will keep probe-level values because they are usually transcript specific.
  #################################################
  # 
  # sampleData <- sampleData %>%
  #   mutate(GENE_SYMBOL = str_replace(PROBE_NAME, "_.*", ""))
  # 
  # sampleDataTable <- data.table(sampleData %>% dplyr::select(-PROBE_NAME))
  # sampleData <- sampleDataTable[, lapply(.SD, sum), by = "GENE_SYMBOL"]
  # 
  # # sampleData <- sampleData %>%
  # #   mutate(GENE_SYMBOL = str_replace(PROBE_NAME, "_.*", "")) %>%
  # #   dplyr::group_by(GENE_SYMBOL) %>%
  # #   dplyr::summarise_if(.predicate = function(x) is.numeric(x),
  # #                       .funs = c("sum")) %>%
  # #   ungroup()
  # sampleData <- as.data.frame(sampleData)
  # row.names(sampleData) <- sampleData$GENE_SYMBOL
  # sampleData <- sampleData[,-1]

  # biospyder %>% select(ENSEMBL_GENE_ID) %>% distinct() %>% count()
  # sampleData <- sampleData[row.names(sampleData) %in% biospyder$ENSEMBL_GENE_ID,] # If aligned to hg38...
  
} else { 
  print("Platform/technology not recognized") 
}

```

The code above (not shown by default) loads user-provided sample meta data (i.e., information about your experiment, also known as colData, or, column data). This also imports the count matrix (i.e., a table of observed counts in which each sample is a column and genes are rows).  

The experimental comparisons of interest to be tested in this report are as follows:  

```{r 'print_contrasts'}
knitr::kable(contrasts,
             row.names = F,
             col.names = c("Group of interest", "Control for comparison"),
             caption = "Contrasts requested for this analysis.") %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "left") %>%
  scroll_box(height = "480px")
```

## Run DESeq2

```{r run-DESeq2, collapse = TRUE, warning = FALSE, message = FALSE, eval = flag}
##########
# DESeq2 #
##########
print(NORM_TYPE) # Name of experiment

# First data clean-up: replace NA & remove samples with total readcount < threshold
initialSampleDataCount <- ncol(sampleData)
sampleData[ is.na(sampleData) ] <- 0 
sampleData <- sampleData[,(colSums(sampleData) > threshold)] # reads required per sample
filteredSampleDataCount <- ncol(sampleData)
# Sometimes extra cleanup may be needed
# colnames(sampleData) <- gsub(pattern = "^0", replacement = "", x = colnames(sampleData))

samples_before <- nrow(DESeqDesignAsRead)

# Sanity check: each sample (row) in the metadata should have a corresponding column in the count data
metadata_in_sampledata <- all(rownames(DESeqDesign) %in% colnames(sampleData))
# Sanity check: each column in the count data should have a corresponding sample (row) in the metadata
sampledata_in_metadata <- all(colnames(sampleData) %in% rownames(DESeqDesign))
# Find samples that were removed because they weren't in metadata
removed <- colnames(sampleData[which(!colnames(sampleData) %in% rownames(DESeqDesign))])
# Reorder the metadata table to correspond to the order of columns in the count data
DESeqDesign <- DESeqDesign[DESeqDesign$original_names %in% colnames(sampleData),]
DESeqDesign <- na.omit(DESeqDesign)
sampleData <- sampleData[,(rownames(DESeqDesign))]
samples_after <- nrow(DESeqDesign)

head(rownames(DESeqDesign))
head(colnames(sampleData)) # Output should match

# Intgroups need to be factors for DESeq2
DESeqDesign[intgroup] <- lapply(DESeqDesign[intgroup], factor)
if (!is.na(params$design)) {
  # If there is a dose column, reorder the experimental group (DESIGN) by dose.
  if (any(grepl(x = colnames(DESeqDesign),
                pattern = "dose",
                ignore.case = T))) {
    doseCol <- grep(x = colnames(DESeqDesign),
                    pattern = "dose",
                    ignore.case = T)
    design_factor_reordered <- factor(DESeqDesign[[params$design]],
                                      levels = unique(DESeqDesign[[params$design]][order(DESeqDesign[[doseCol]])]),
                                      ordered = FALSE)
    DESeqDesign[[params$design]] <- design_factor_reordered
  } else {
    DESeqDesign[params$design] <- factor(DESeqDesign[, params$design])
  }
}

if (file.exists(file.path(paths$RData, "dds.RData")) & is.na(params$group_facet) & params$use_cached_RData == TRUE) {
  print(paste("Already found DESeq2 object from previous run; loading from disk."))
  load(file.path(paths$RData, "dds.RData"))
  if (!identical(as.data.frame(round(counts(dds))),
                 round(sampleData), 0)) {
    print("Not identical")
    }
  } else {
    if (file.exists(file.path(paths$RData,
                             paste0("dds_",
                                    paste(params$group_filter, collapse = "_"),
                                    ".RData"))) & 
       !is.na(params$group_facet) & 
       params$use_cached_RData == TRUE) {
      load(file = file.path(paths$RData,
                          paste0("dds_",
                                 paste(params$group_filter, collapse = "_"),
                                 ".RData")))
      } else {
      current_design <- formula(paste0(paste0("~"),
                                       paste0(intgroup, collapse = " + ")))
      dds <- DESeqDataSetFromMatrix(countData = round(sampleData),
                                    colData   = as.data.frame(DESeqDesign),
                                    design    = current_design)
      dds <- dds[, rownames(DESeqDesign)]
      dds <- dds[rowSums(counts(dds)) > 1]
      dds <- DESeq(dds, parallel = TRUE, BPPARAM = MulticoreParam(params$cpus))
      if (is.na(params$group_facet)) {
        save(dds,
             file = file.path(paths$RData, "dds.RData"))
      } else {
        save(dds,
             file = file.path(paths$RData,
                            paste0("dds_",
                                   paste(params$group_filter, collapse = "_"),
                                   ".RData")))
      }
    }
}

# ### REMOVE if done above from scratch
# keep <- grep(colData(dds)$chemical, pattern = "cells", invert = T, ignore.case = T)
# dds <- dds[, keep]
# dds <- dds[row.names(dds) %in% biospyder$ENSEMBL_GENE_ID,]

# Another sanity check to make sure the object looks correct
resultsNames(dds)
head(colData(dds))
head(assay(dds))
head(rowRanges(dds))
str(counts(dds))

# Make regularized log object for later plotting
# rld <- tryCatch(rlog(dds), error = function(e) { rlog(dds, fitType = 'mean') })
# Use vst for hundreds of samples!

if (!is.na(params$nuisance)) {
  rld <- vst(dds, blind = F) # Should this be blind?
  mat <- assay(rld)
  condition <- formula(paste0(paste0("~"),
                              paste0(intgroup[!intgroup %in% params$nuisance],
                                     collapse = " + ")))
  mm <- model.matrix(condition, colData(rld))
  mat <- limma::removeBatchEffect(mat, batch = rld[[params$nuisance]], design = mm)
  assay(rld) <- mat
} else {
  rld <- vst(dds, blind = F) # Should this be blind?
}

# rld <- rlog(dds, blind = F)
```

The code above (not shown by default) uses DESeq2 `r packageVersion("DESeq2")` to test for deferentially abundant genes within the `r Platform` data.  

# Quality control

A study-wide quality control script removes samples that cluster at a distance greater than a specified cutoff, samples that are noisy, and samples that do not have sufficient aligned reads. **More details are available in a separate report.**  

Within this report, gene/probe-level quality control is performed below ("Run pairwise comparisons" section).

Prior to running DESeq2, the data was filtered to remove samples that do not have $`r threshold`$ reads per sample.  

The metadata used for this report included `r samples_before` samples.  

The count matrix initially included `r initialSampleDataCount` samples (including any reference material samples). After removing samples with less than $`r threshold`$ reads, `r filteredSampleDataCount` samples were left. It is `r metadata_in_sampledata` that all the samples provided in the metadata table were also identified in the count matrix. It is `r sampledata_in_metadata` that all the samples in the count matrix were also identified in the metadata table.  

There were `r samples_after` samples passing all thresholds. **The samples excluded from analysis are shown in the table in the section titled "Sample data", below, along with complete sample metadata for the experiment.**    

## Sample data (metadata about your experiment) {.tabset .tabset-fade}

**Some samples may already have been removed in study-wide QC.**

### Samples used in this report

This table shows the final list of samples that were used in the data analysis (as well as the corresponding sample information, e.g., to which experimental group samples belong).  

```{r print_metadata_used}
# Conditionally sort by dose if that column name is present
if (any(grepl(x = colnames(DESeqDesign), pattern = "dose", ignore.case = T))) {
  doseCol <- grep(x = colnames(DESeqDesign), pattern = "dose", ignore.case = T)
  knitr::kable(DESeqDesign %>%
                 group_by(!!sym(params$design)) %>%
                 arrange(!!sym(colnames(DESeqDesign)[doseCol])),
             row.names =  F,
             caption = "Samples and corresponding experimental conditions used in this report") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")
} else {
  knitr::kable(DESeqDesign %>% arrange(!!sym(params$design)),
             row.names =  F,
             caption = "Samples and corresponding experimental conditions used in this report") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")
}

knitr::kable(DESeqDesign %>% group_by(!!sym(params$design)) %>% tally(),
             row.names = F,
             col.names = c("Experimental group", "Number of samples in group"),
             caption = "Number of samples in each experimental group used in this report") %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "left") %>%
  scroll_box(width = "100%", height = "480px")
```

### Samples removed from this report

This table shows the samples that were removed from this analysis.

```{r print_metadata_removed}
# Conditionally sort by dose if that column name is present?
knitr::kable(DESeqDesignAsRead %>% arrange(!!sym(params$design)) %>% dplyr::filter(!original_names %in% DESeqDesign$original_names),
             row.names =  F,
             caption = "Samples removed from analysis") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

knitr::kable(DESeqDesignAsRead %>%
               dplyr::filter(original_names %in% removed) %>%
               group_by(!!sym(params$design)) %>% tally(),
             row.names = F,
             col.names = c("Experimental group", "Number of samples removed"),
             caption = "Number of samples removed from each experimental group") %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "left") %>%
  scroll_box(width = "100%", height = "480px")
```

### Original sample metadata as provided

This table shows the original sample metadata including any samples that were removed within this report.  **Note that some samples may already have been removed in study-wide QC before loading this metadata.**

```{r print_metadata_provided}
# Conditionally sort by dose if that column name is present?
knitr::kable(DESeqDesignAsRead %>% arrange(!!sym(params$design)),
             row.names =  F,
             caption = "Complete sample metadata as provided to the Genomics Laboratory.") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>%
  scroll_box(width = "100%", height = "480px")

knitr::kable(DESeqDesignAsRead %>% group_by(!!sym(params$design)) %>% tally(),
             row.names = F,
             col.names = c("Experimental group", "Number of samples as provided"),
             caption = "Number of samples and experimental groups as provided") %>%
  kable_styling(bootstrap_options = "striped", full_width = F, position = "left") %>%
  scroll_box(width = "100%", height = "480px")
```

## Read count plots

This section shows the number of reads obtained for sample groups included in the report.  

```{r read-counts, warning = FALSE}
reads_df = data.frame(reads = round(colSums(sampleData), 3))
reads_df$original_names <- row.names(reads_df)

reads_df_annotated <- dplyr::left_join(DESeqDesign, reads_df)

ggplot(reads_df,
       aes(x = !!sym(params$design),
           y = reads)) +
  geom_boxplot() +
  geom_jitter(size = 0.3, alpha = 0.5, color = "blue") +
  coord_flip() +
  expand_limits(y = 0) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  scale_y_continuous(labels = function(x) format(x, scientific = TRUE))

```

# Testing for differentially expressed genes

## Load R-ODAF functions

```{r 'define-ODAF-functions', collapse = TRUE, warning = FALSE, message = FALSE}
###################################################################################
#DEFINE FUNCTIONS
###################################################################################
plot.barplots <- function(samples, b) {
  color <- NULL
  for (h in 1:ncol(norm_data)) {
    if (substring(colnames(norm_data)[h], 1, 3) == substring(condition2, 1, 3)) { color <- c(color, "red3") } else {color <- c(color, "darkgrey")}
  }
  fileNamePlot <- paste0(b, row.names(samples)[b], ".png")
  pseudoTitle <- paste0(row.names(samples)[b], "_pAdj:", samples[b, "padj"])
  
  png(file = paste(fileNamePlot, sep = "/"), width = 1200, height = 700, pointsize = 20)
  par(mar = c(8, 4, 3, 1))
  barplot(as.numeric(norm_data[row.names(samples)[b],]), las = 2, col = color, main = pseudoTitle, cex.names = 0.5,  cex.axis = 0.8, names.arg = colnames(norm_data)) 
  dev.off()
} #plot.barplots function done

###################################################################################
draw.barplots <- function(samples, top_bottom, NUM){
  if (nrow(samples) == 0) {
    #print("no genes to plot") 
  } else { 
    if (top_bottom == "top") {
      #print(paste0("drawing Top ", NUM, " plots"))
      if (nrow(samples) <= NUM) { 
        for (b in 1:nrow(samples)) {plot.barplots(samples, b)}
      }
      
      if (nrow(samples) > NUM) { 
        for (b in 1:NUM) {plot.barplots(samples, b)}	
      }	
    }
    
    if (top_bottom == "bottom") {
      #print(paste0("drawing Bottom", NUM, " plots"))
      if (nrow(samples) <= NUM) { 
        for (b in 1:nrow(samples)) {plot.barplots(samples, b)}
      }
      if (nrow(samples) > NUM) { 
        for (b in ((nrow(samples) - NUM + 1):nrow(samples))) {plot.barplots(DEsamples, b)}
      }
    }}
} #draw.barplots function done

###################################################################################
###################################################################################
```

The code above (not shown by default) loads in plotting functions specific to the Omics Data Analysis Frameworks for Regulatory application (R-ODAF) template. More information on the R-ODAF framework is available [here](https://github.com/MCTverheijen/R-ODAF).

## Run pairwise comparisons

```{r 'run_ODAF_code', collapse = TRUE, warning = FALSE, message = FALSE, eval = flag}

# Initial setup for DESeq2 contrasts
cooks   <- F # Sometimes may need consideration
resList <- list()
resListAll <- list()
Counts  <- counts(dds, normalized = TRUE)
CPMdds  <- cpm(counts(dds, normalized = TRUE))

for (x in 1:nrow(contrasts)) {	## for all comparisons to be done	
	
  condition1 <- contrasts[x, 2]
	condition2 <- contrasts[x, 1]
	print(paste(condition2, " vs ", condition1, ":", NORM_TYPE))
	
	DE_Design <- matrix(data = NA, ncol = 2)
	DE_Design <- as.matrix(DESeqDesign[DESeqDesign[, DESIGN] %in% c(condition1,
	                                                               condition2),])
	samples   <- sampleData[, rownames(DE_Design)]
	colnames(samples) <- NULL
	
	###########

	print(paste0("Filtering genes: 75% of at least 1 group need to be above ", MinCount, " CPM"))
	SampPerGroup <- table(DE_Design[, DESIGN])
  Counts <- counts(dds, normalized = TRUE)
  CPMdds <- cpm(counts(dds, normalized = TRUE))
  		
  Filter <- matrix(data = NA, ncol = 3, nrow = nrow(Counts))
  rownames(Filter) <- rownames(Counts)
  colnames(Filter) <- c("Low", "quantile", "spike")
  		
  # Apply the "Relevance" condition
  	
  	for (gene in 1:nrow(dds)) {
  		CountsPass <- NULL
  			for (group in 1:length(SampPerGroup)) { 
  				sampleCols <- grep(dimnames(SampPerGroup)[[1]][group], DE_Design[, DESIGN])
  				Check <- sum(CPMdds[gene,sampleCols] >= MinCount) >= 0.75 * SampPerGroup[group]
  				CountsPass <- c(CountsPass, Check)
  		}
  	if (sum(CountsPass) > 0) {Filter[gene, 1] <- 1 }	else {Filter[gene,1] <- 0 }
  	}

  compte <- Counts[Filter[,1] == 1,]
  Filter <- Filter[rownames(Filter) %in% rownames(compte), , drop = F]

  print(paste0("Relevance filtering removed ", nrow(dds) - nrow(Filter),
              " genes from the ", nrow(dds)," assessed. ",
              nrow(Filter), " genes remaining"))
	
  print("Obtaining the DESeq2 results")

	currentContrast <- c(DESIGN, condition2, condition1)
	res <- DESeq2::results(dds[rownames(compte),],
	               parallel = TRUE,
	               BPPARAM = MulticoreParam(params$cpus),
	               contrast = currentContrast,
	               pAdjustMethod = 'fdr',
	               cooksCutoff = cooks) # If Cooks cutoff disabled - manually inspect.
	res <- lfcShrink(dds = dds[rownames(compte),],
	                 contrast = currentContrast,
	                 res = res,
	                 type = "ashr")
	
	# Make new directory for the ODAF-specific files
	ODAFdir <- file.path(paths$DEG_output, "R-ODAF")
	if (!dir.exists(ODAFdir)) {dir.create(ODAFdir, recursive = TRUE)}
	FileName <- paste(NORM_TYPE, condition2, "vs", condition1, "FDR", pAdjValue, sep = "_")
	
	# Create output tables
	norm_data <<- counts(dds[rownames(compte)], normalized = TRUE)
	DEsamples <<- subset(res, res$padj < pAdjValue)
	DECounts <- compte[rownames(compte) %in% rownames(DEsamples), , drop = F]
	Filter <- Filter[rownames(Filter) %in% rownames(DECounts), , drop = F]
	
	print("Check median against third quantile" )
	print("AND")
	print("Check the presence of a spike" )

	for (gene in 1:nrow(DECounts)) {
	  # Check the median against third quantile
		quantilePass <- NULL
		sampleColsg1 <- grep(dimnames(SampPerGroup)[[1]][1],DE_Design[,DESIGN])
		sampleColsg2 <- grep(dimnames(SampPerGroup)[[1]][2],DE_Design[,DESIGN])
		
	  Check <- median(as.numeric(DECounts[gene, sampleColsg1])) > quantile(DECounts[gene, sampleColsg2], 0.75)[[1]]
		quantilePass <- c(quantilePass, Check)
		
		Check <- median(as.numeric(DECounts[gene, sampleColsg2])) > quantile(DECounts[gene, sampleColsg1], 0.75)[[1]]
		quantilePass <- c(quantilePass, Check)
			
		if (sum(quantilePass) > 0) {
		  Filter[gene, 2] <- 1
		}	else {
		  Filter[gene, 2] <- 0
		}

		# Check for spikes 
		spikePass <- NULL
		for (group in 1:length(SampPerGroup)) {
				sampleCols <- grep(dimnames(SampPerGroup)[[1]][group], DE_Design[ ,DESIGN])
			if (max(DECounts[gene,sampleCols]) == 0) {Check <- FALSE} else {
				Check <- (max(DECounts[gene, sampleCols]) / sum(DECounts[gene, sampleCols])) >= 1.4 * (SampPerGroup[group])^(-0.66)
				spikePass <- c(spikePass, Check)
			}
		}
		if (sum(spikePass) > 1) {
		  Filter[gene, 3] <- 0
		}	else {
		  Filter[gene, 3] <- 1
		}
	}
	
	# Extract the final list of DEGs

	DECounts_real <- DEsamples[rowSums(Filter) == 3 ,]
	DECounts_no_quant <- DEsamples[Filter[, 2] == 0 ,]
	DECounts_spike <- DEsamples[Filter[, 3] == 0 ,]
	
  print(paste0("A total of ", nrow(DECounts_real),
              " DEGs were selected, after ", nrow(DECounts_no_quant),
              " genes(s) removed by the quantile rule and ", nrow(DECounts_spike),
              " gene(s) with a spike"))

 # Save the normalized counts and the list of DEGs
	write.table(norm_data,
	            file = paste0(ODAFdir, FileName, "_Norm_Data.txt"),
	            sep = "\t",
	            quote = FALSE)
	write.table(DEsamples,
	            file = paste0(ODAFdir, FileName, "_DEG_table.txt"),
	            sep = "\t",
	            quote = FALSE)
	write.table(DECounts_no_quant,
	            file = paste0(ODAFdir,FileName, "_failed_quantile_table.txt"),
	            sep = "\t",
	            quote = FALSE)
	write.table(DECounts_spike,
	            file = paste0(ODAFdir, FileName, "_DEspikes_table.txt"),
	            sep = "\t",
	            quote = FALSE)

	print("DESeq2 Done")
	
	colnames(norm_data) <- colData(dds)[, DESIGN]

	resListAll[[x]] <- res
	resList[[x]] <- DECounts_real
	
	if (R_ODAF_plots == TRUE) {
  	print("creating Read count Plots")
  	# top DEGs
  	plotdir <- file.path(paths$DEG_output, "plots")
  	if (!dir.exists(plotdir)) {dir.create(plotdir, recursive = TRUE)}
  	barplot.dir <- file.path(paths$DEG_output, "plots", "/barplot_genes/")
  	if (!dir.exists(barplot.dir)) {dir.create(barplot.dir, recursive = TRUE)}
  
  	TOPbarplot.dir <- file.path(barplot.dir, "Top_DEGs")
  	if (!dir.exists(TOPbarplot.dir)) {dir.create(TOPbarplot.dir, recursive = TRUE)}
  	setwd(TOPbarplot.dir)
  	draw.barplots(DEsamples, "top", 20) # (DEsamples, top_bottom, NUM)
  	print("Top 20 DEG plots done")
  
  	# Spurious spikes
  	SPIKEbarplot.dir <- file.path(barplot.dir, "DE_Spurious_spikes")
  	if (!dir.exists(SPIKEbarplot.dir)) {dir.create(SPIKEbarplot.dir, recursive = TRUE)}
  	setwd(SPIKEbarplot.dir)
  	draw.barplots(DEspikes, "top", nrow(DEspikes)) # (DEsamples, top_bottom, NUM)
  	print("All DE_Spurious_spike plots done")
	}
}
```

The code above (not shown by default) runs the R-ODAF spurious spike detection and outputs the DESeq Results objects as a list for each contrast. As specified by the R-ODAF guidelines, 75% of at least 1 group need to be above `r MinCount` CPM and spurious spikes were removed in which Max-Median > Sum/(Rep + 1).  

The log2FoldChange shrinkage procedure used was `r resList[[1]]@priorInfo$type`. An alpha of `r resList[[1]]@metadata$alpha` was used to extract raw results, which are reported as the `r gsub(mcols(x=resList[[1]])$description[[4]],pattern=":.*",replacement="")`. To account for multiple testing, `r mcols(resList[[1]])$description[5]` are reported. Cook's cutoff was set to `r cooks` in this analysis.  


## Create annotation tables

```{r}
################################################################################
# Annotations
################################################################################

#listEnsembl()
#listMarts()
#listDatasets(useMart('ensembl'))
ensembl <- useMart("ensembl", dataset = ensembl_species,  host = "useast.ensembl.org")
#listAttributes(ensembl)
#listFilters(ensembl)


genes <- lapply(resListAll,
                function(x) row.names(as.data.frame(x)))
genes <- unlist(genes) %>% unique()
#genes_all <- unique(row.names(assay(dds)))

if (file.exists(file.path(paths$RData, "id_table.Rdata"))) {
  print(paste("Already found ID table; skipping biomaRt query and loading from disk."))
  load(file.path(paths$RData, "id_table.Rdata"))
  id_table <- id_table_entrez[, 1:3]
  } else {
  id_table_entrez <- getBM(filters = biomart_filter,
                        attributes = c(biomart_filter,
                                      species_gene_symbol,
                                      "description",
                                      "entrezgene_id",
                                      "entrezgene_accession"),
                        values = genes,
                        mart = ensembl)
  save(id_table_entrez, file = file.path(paths$RData, "id_table.Rdata"))
  id_table <- id_table_entrez[, 1:3]
}
```


## Create summary tables

```{r create_tables, collapse = TRUE, warning = FALSE, message = FALSE, eval = flag}
################################################################################
# Table modifications
################################################################################

annotate_deseq_table <- function(deseq_results_list, filter_results = F) {
  x <- deseq_results_list
  annotated_results <- list()
  for (i in 1:length(x)) {
    print(i)
    deg_table <- x[[i]]
    # Add taxonomy
    if (nrow(deg_table) == 0) {
      next
    } else {
      deg_table <- cbind(genes = row.names(x[[i]]),
                      as(deg_table, "data.frame"),
                      contrast = gsub(pattern = paste0("log2.*", DESIGN, "\ "),
                                      replacement =  "",
                                      x = x[[i]]@elementMetadata[[2]][2]))
      names(deg_table)[names(deg_table) == "genes"] <- biomart_filter
      deg_table <- dplyr::left_join(deg_table,
                                    id_table,
                                    by = biomart_filter)
      deg_table <- dplyr::mutate(deg_table, linearFoldChange = ifelse(log2FoldChange > 0,
                                                              2 ^ log2FoldChange,
                                                              -1 / (2 ^ log2FoldChange)))
      deg_table <- deg_table[, c(1, 8, 9, 2, 3, 10, 4:7)]
      ## FILTERS ##
      if (filter_results == T) {
      deg_table <- deg_table[!is.na(deg_table$padj) & deg_table$padj < alpha & abs(deg_table$linearFoldChange) > linear_fc_filter, ]
      }
      annotated_results[[i]] <- deg_table %>% dplyr::distinct()
    }
  }
  y <- rbindlist(annotated_results)
  return(y)
}

# Write dataframe of all results
# Significant only
significantResults <- annotate_deseq_table(resList, filter_results = T)
significantResults <- significantResults %>% dplyr::mutate(contrast = as_factor(contrast))
# All results including non-significant
allResults <- annotate_deseq_table(resListAll, filter_results = F)
allResults <- allResults %>% dplyr::mutate(contrast = as_factor(contrast))

# All Results for Plotting
# Replace NA gene symbols with ensembl ID
allResults[[species_gene_symbol]][is.na(allResults[[species_gene_symbol]])] <- allResults[[biomart_filter]][is.na(allResults[[species_gene_symbol]])] 
# Replace blank gene symbols with ensembl ID
allResults[[species_gene_symbol]][allResults[[species_gene_symbol]] == ""]  <- allResults[[biomart_filter]][allResults[[species_gene_symbol]] == ""]

IPA <- allResults %>% 
  distinct() %>% 
  pivot_wider(names_from = contrast, values_from = c(log2FoldChange, linearFoldChange, lfcSE, pvalue, padj))
allResults$padj[allResults$padj == 0] <- 10^-100 # For plotting purposes!
allResultsOrdered_logFC_filter <- dplyr::filter(allResults, abs(linearFoldChange) > 1.5) %>%
  arrange(-abs(linearFoldChange))
allResultsOrdered_logFC_filter <- dplyr::filter(allResultsOrdered_logFC_filter, padj < alpha)
res.df <- allResultsOrdered_logFC_filter

degTable <- significantResults %>%
  dplyr::group_by(contrast) %>%
  dplyr::count()

lengths <- lapply(resList, nrow)
longest <- which.max(lengths)
summaryTable <- data.frame( genes = row.names(resList[[longest]]),
                            baseMean = resList[[longest]]$baseMean )
names(summaryTable)[names(summaryTable) == "genes"] <- biomart_filter
contrastsInSummary <- vector()
for (i in 1:length(resList)) {
  print(i)
  n <- resList[[i]]@elementMetadata[[2]][2]
  n <- gsub(pattern = paste0("log2\ fold\ change\ \\(MMSE\\):\ ", DESIGN), # Might need to be MLE for some tests
            replacement =  paste0("log2 Fold Change"),
            x = resList[[i]]@elementMetadata[[2]][2])
  p <- gsub(pattern = paste0("log2\ fold\ change\ \\(MMSE\\):\ ", DESIGN),
           replacement =  resList[[i]]@elementMetadata[[2]][6], # Not used?
           x = resList[[i]]@elementMetadata[[2]][2])
  q <- gsub(pattern = paste0("log2\ fold\ change\ \\(MMSE\\):\ ", DESIGN, "\ "),
            replacement =  "",
            x = resList[[i]]@elementMetadata[[2]][2])
  message(n)
  message(p)
  message(q)
  toJoin <- as.data.frame(resList[[i]])
  setDT(toJoin, keep.rownames = T)[]
  setnames(toJoin, 1, biomart_filter)
  toJoin <- mutate(toJoin, linearFoldChange = ifelse(log2FoldChange > 0,
                                                2 ^ log2FoldChange,
                                                -1 / (2 ^ log2FoldChange)))
  toJoin <- toJoin[, c(1:3, 7, 4:6)]
  summaryTable <- dplyr::left_join(summaryTable, dplyr::select(toJoin, !c(baseMean, pvalue, lfcSE)), by = biomart_filter)

  names(summaryTable)[[ncol(summaryTable) - 2]] <- paste0("log2FoldChange_", i)
  names(summaryTable)[[ncol(summaryTable) - 1]] <- paste0("linearFoldChange_", i)
  names(summaryTable)[[ncol(summaryTable)]] <- paste0("FDR_", i)
  contrastsInSummary[i] <- q
  print(summary(resList[[i]], pAdjValue))
}

maxFCs <- allResults %>%
  dplyr::group_by(!!sym(biomart_filter)) %>%
  dplyr::filter(abs(linearFoldChange) == max(abs(linearFoldChange))) %>%
  dplyr::ungroup() %>%
  dplyr::select(!!sym(biomart_filter), linearFoldChange)

minPvals <- allResults %>%
  group_by(!!sym(biomart_filter)) %>%
  dplyr::filter(padj == min(padj)) %>%
  dplyr::ungroup() %>%
  dplyr::select(!!sym(biomart_filter), padj)

summaryTable <- summaryTable %>%
  left_join(id_table, by = biomart_filter) %>%
  left_join(maxFCs, by = biomart_filter) %>%
  left_join(minPvals, by = biomart_filter) %>%
  dplyr::rename(maxFoldChange = linearFoldChange,
         minFDR_pval = padj) %>%
  dplyr::distinct() %>%
  mutate(maxFoldChange = abs(maxFoldChange)) # This eliminates the direction of change: this way it's easy to sort.

numColsToPrepend <- ncol(summaryTable) - 3*length(resList) - 2 # Number of columns per contrast = 3. Subtract two for the baseMean and genes columns.
colPositionsToPrependSTART <- ncol(summaryTable) - numColsToPrepend + 1
colPositionsOfData <- ncol(summaryTable) - numColsToPrepend
summaryTable <- summaryTable[, c(1,
                                colPositionsToPrependSTART:ncol(summaryTable),
                                2:colPositionsOfData)]
summaryTable <- summaryTable %>% dplyr::distinct() # Just in case duplicates snuck by

CPMddsDF <- data.frame(genes = row.names(CPMdds), CPMdds, check.names = F)
CPMddsDF <- dplyr::left_join(CPMddsDF, id_table, by = c("genes" = biomart_filter))
numColsToPrepend <- ncol(CPMddsDF) - ncol(CPMdds) - 1
colPositionsToPrependSTART <- ncol(CPMddsDF) - numColsToPrepend + 1
colPositionsOfData <- ncol(CPMddsDF) - numColsToPrepend
CPMddsDF <- CPMddsDF[, c(1, colPositionsToPrependSTART:ncol(CPMddsDF), 2:colPositionsOfData)]


```

The code above (not shown by default) generates tables summarizing the differentially expressed genes (DEGs).  A linear fold change cutoff of `r linear_fc_filter` and adjusted p-value of `r alpha` was used to filter the results.

Here is the number of DEGs in each group:

```{r 'display_DEG_summary'}
kable(degTable,
      caption = "Number of differentially expressed genes across each contrast") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

```{r 'write-tables', collapse = TRUE, warning = FALSE, message = FALSE, eval = flag}
#######################################
### Write results table from DESeq2
#######################################
write.table(allResults,
            file = file.path(paths$DEG_output, NORM_TYPE,"-DESeq_output_ALL.txt"),
            quote = F, sep = '\t', col.names = NA)
write.table(significantResults,
            file = paste0(paths$DEG_output, NORM_TYPE, "-DESeq_output_significant.txt"),
            quote = F, sep = '\t', col.names = NA)
write.table(summaryTable,
            file = paste0(paths$DEG_output, NORM_TYPE, "-DESeq_output_all_genes.txt"),
            quote = F, sep = '\t', col.names = NA)
write.table(CPMddsDF,
            file = paste0(paths$DEG_output, NORM_TYPE, "-Per_sample_CPM.txt"),
            quote = F, sep = '\t', col.names = NA)
write.table(Counts,
            file = paste0(paths$DEG_output, NORM_TYPE, "-Per_sample_normalized_counts.txt"),
            quote = F, sep = '\t', col.names = NA)
```

The code above (not shown by default) writes text files for each DEG summary type.

# Write data

```{r 'write-excel-worksheets', collapse = TRUE, warning = FALSE, message = FALSE, eval = flag}
#######################################
### Write results above but in Excel
#######################################
### Global options
options("openxlsx.borderColour" = "#4F80BD")
options("openxlsx.borderStyle" = "thin")
options("openxlsx.maxWidth" = 50)
hs1 <- createStyle(textDecoration = "Bold",
                   border = "Bottom",
                   fontColour = "black")
hs2 <- createStyle(textDecoration = "Bold",
                   border = c("top", "bottom", "left", "right"),
                   fontColour = "black",
                   fgFill = "#C5D9F1")

### Summary results - one gene per line, columns are contrasts
wb1 <- createWorkbook()
modifyBaseFont(wb1, fontSize = 10, fontName = "Arial Narrow")
addWorksheet(wb1, "DESeq_results_per_gene")
for (j in 1:length(contrastsInSummary)) {
  myStartcol = 7 + ((j - 1) * 3)
  myEndcol = 9 + ((j - 1) * 3)
  mergeCells(wb1,
             sheet = 1,
             cols = myStartcol:myEndcol,
             rows = 1)
  writeData(
    wb1,
    sheet = 1,
    x = contrastsInSummary[j],
    startCol = myStartcol,
    startRow = 1)
}
conditionalFormatting(wb1,
                      sheet = 1,
                      rows = 1,
                      cols = 1:ncol(summaryTable),
                      type = "contains",
                      rule = "",
                      style = hs2)
freezePane(wb1, sheet = 1, firstActiveRow = 3, firstActiveCol = 4)
writeDataTable(wb1,
               sheet = 1,
               startRow = 2,
               x = summaryTable,
               colNames = TRUE,
               rowNames = F,
               tableStyle = "none",
               headerStyle = hs1,
               keepNA = T,
               na.string = "NA")
setColWidths(wb1, sheet = 1, cols = 1:6, widths = "auto") # This is hard-coded, so prone to error; will only impact auto adjustment of col widths.
setColWidths(wb1, sheet = 1, cols = 7:ncol(summaryTable), widths = 13) # This is hard-coded, so prone to error; will only impact auto adjustment of col widths.
fname1 <- file.path(paths$DEG_output, "1.", NORM_TYPE, "-DESeq_by_gene.xlsx")
saveWorkbook(wb1, fname1, overwrite = TRUE)

### All results in one table
wb2 <- createWorkbook()
modifyBaseFont(wb2, fontSize = 10, fontName = "Arial Narrow")
addWorksheet(wb2, paste0("FDR", pAdjValue, ".Linear.FC.", linear_fc_filter))
freezePane(wb2, sheet = 1, firstRow = TRUE, firstActiveCol = 4)
writeDataTable(wb2,
               sheet = 1,
               x = significantResults,
               colNames = TRUE,
               rowNames = F,
               tableStyle = "none",
               headerStyle = hs1,
               keepNA = T,
               na.string = "NA")
setColWidths(wb2, sheet = 1, cols = 1:ncol(significantResults), widths = "auto")
addWorksheet(wb2, "DESeq_all_results")
freezePane(wb2, sheet = 2, firstRow = TRUE, firstActiveCol = 4)
writeDataTable(wb2,
               sheet = 2,
               x = allResults,
               colNames = TRUE,
               rowNames = F,
               tableStyle = "none",
               headerStyle = hs1,
               keepNA = T,
               na.string = "NA")
setColWidths(wb2, sheet = 2, cols = 1:ncol(allResults), widths = "auto")
fname2 <- file.path(paths$DEG_output, "2.", NORM_TYPE, "-DESeq_all.xlsx")
saveWorkbook(wb2, fname2, overwrite = TRUE)

### All results with different tabs for each contrast
wb3 <- createWorkbook()
modifyBaseFont(wb3, fontSize = 10, fontName = "Arial Narrow")

short_contrast_names <- stringr::str_trunc(short_contrast_names,
                                           31,
                                           side = "right",
                                           ellipsis = "")

for (i in 1:length(levels(factor(allResults$contrast)))) {
  print(i)
  dataToWrite <- allResults[allResults$contrast == levels(factor(allResults$contrast))[i],]
  addWorksheet(wb3, short_contrast_names[i])
  freezePane(wb3, sheet = i, firstRow = TRUE, firstActiveCol = 4)
  writeDataTable(wb3,
                 sheet = i,
                 x = dataToWrite,
                 colNames = TRUE,
                 rowNames = F,
                 tableStyle = "none",
                 headerStyle = hs1,
                 keepNA = T,
                 na.string = "NA")
  setColWidths(wb3, sheet = i, cols = 1:ncol(dataToWrite), widths = "auto")
}
fname3 <- file.path(paths$DEG_output, "3.", NORM_TYPE, "-DESeq_by_contrast.xlsx")
saveWorkbook(wb3, fname3, overwrite = TRUE)

### CPM
wb4 <- createWorkbook()
modifyBaseFont(wb4, fontSize = 10, fontName = "Arial Narrow")
addWorksheet(wb4, "Counts per million")
freezePane(wb4, sheet = 1, firstRow = TRUE, firstActiveCol = 4)
writeDataTable(wb4,
               sheet = 1,
               x = as.data.frame(CPMddsDF),
               colNames = TRUE,
               rowNames = F,
               tableStyle = "none",
               headerStyle = hs1,
               keepNA = T,
               na.string = "NA")
setColWidths(wb4, sheet = 1, cols = 1:ncol(CPMddsDF), widths = "auto")
fname4 <- file.path(paths$DEG_output, "4.", NORM_TYPE, "-CPM.xlsx")
saveWorkbook(wb4, fname4, overwrite = TRUE)

### IPA
wb5 <- createWorkbook()
modifyBaseFont(wb5, fontSize = 10, fontName = "Arial Narrow")
addWorksheet(wb5, "For IPA upload")
freezePane(wb5, sheet = 1, firstRow = TRUE, firstActiveCol = 5)
writeDataTable(wb5,
               sheet = 1,
               x = as.data.frame(IPA),
               colNames = TRUE,
               rowNames = F,
               tableStyle = "none",
               headerStyle = hs1,
               keepNA = T,
               na.string = "NA")
setColWidths(wb5, sheet = 1, cols = 1:ncol(IPA), widths = "auto")
fname5 <- file.path(paths$DEG_output, "5.", NORM_TYPE, "-IPA.xlsx")
saveWorkbook(wb5, fname5, overwrite = TRUE)

```

The code above (not shown by default) writes Excel workbooks and text files of DEG lists.

These files should be provided to you as a separate zip file.

```{r 'embed_excel_files', eval = embedFiles}
# setwd(paths$DEG_output)
# xfun::embed_files(list.files(".", "[.](xlsx)$"), name = "RNASeq_Spreadsheets.zip")
```

```{r 'embed_text_files', eval = embedFiles}
# setwd(paths$DEG_output)
# xfun::embed_files(list.files(".", "[.](txt)$"), name = "RNASeq_Text_Files.zip")
```

```{r 'embed_ODAF_files', eval = embedFiles}
# setwd(paths$DEG_output)
# xfun::embed_dir(ODAFdir, name = "RNASeq_R-ODAF_text_files.zip")
```


# PCA Plots {.tabset .tabset-fade}

## All `r nrow(assay(rld))` genes, before filtering

```{r 'PCA'}
## Perform PCA analysis and make plot
plotPCA(rld, intgroup = params$design, ntop = nrow(assay(rld))) + 
  theme(legend.position = "bottom") +
  guides(color = guide_legend(ncol = 3))
## Get percent of variance explained
data_pca <- plotPCA(rld, intgroup = params$design, ntop = nrow(assay(rld)), returnData = TRUE)
percentVar <- round(100 * attr(data_pca, "percentVar"))

```

This plot shows the first two principal components that explain the variability in the data using the regularized log count data. If you are unfamiliar with principal component analysis, you might want to check the [Wikipedia entry](https://en.wikipedia.org/wiki/Principal_component_analysis) or this [interactive explanation](http://setosa.io/ev/principal-component-analysis/). In this case, the first and second principal component explain `r percentVar[1]` and `r percentVar[2]` percent of the variance respectively.

## DEGs only (`r length(unique(significantResults[[biomart_filter]]))` genes)

```{r 'PCA_DEGs'}
rld_degs <- rld[row.names(assay(rld)) %in% significantResults[[biomart_filter]],]  # d[ d[[THECOLUMN]] == someValue , ]

## Perform PCA analysis and make plot
plotPCA(rld_degs, intgroup = params$design, ntop = nrow(rld_degs)) +
  theme(legend.position = "bottom") +
  guides(color = guide_legend(ncol = 3))
PCAplotDEGs <- plotPCA(rld_degs, intgroup = params$design, ntop = nrow(rld_degs), returnData = TRUE)

## Get percent of variance explained
percentVarDEGs <- round(100 * attr(PCAplotDEGs, "percentVar"))
```

This plot shows the principal components analysis limited to all DEGs. In this case, the first and second principal component explain `r percentVarDEGs[1]` and `r percentVarDEGs[2]` percent of the variance respectively.

## Top `r nBest` most variable genes only

```{r 'PCA_top'}

# Run this code only once for both the PCA and clustering analysis
rv = rowVars(assay(rld))
select = order(rv, decreasing = TRUE)[1:nBest]

rld_top <- rld[select,]

## Perform PCA analysis and make plot
plotPCA(rld_top, intgroup = params$design, ntop = nrow(rld_top)) + 
  theme(legend.position = "bottom") +
  guides(color = guide_legend(ncol = 3))

## Get percent of variance explained
data_pcaTop <- plotPCA(rld_top, intgroup = params$design, returnData = TRUE, ntop = nrow(rld_top))
percentVarTop <- round(100 * attr(data_pcaTop, "percentVar"))
```

This plot shows the principal components analysis limited to all DEGs. In this case, the first and second principal component explain `r percentVarTop[1]` and `r percentVarTop[2]` percent of the variance respectively.

# Sample-to-sample distances {.tabset .tabset-fade}

## All `r nrow(assay(rld))` genes, before filtering

```{r 'sampleDist', fig.width = 10, fig.height = 10}
## Obtain the sample euclidean distances
sampleDists <- stats::dist(t(assay(rld)))
sampleDists <- dendextend::sort_dist_mat(sampleDists)
sampleDistMatrix <- as.matrix(sampleDists)

## Add names based on intgroup
# rownames(sampleDistMatrix) <- apply(as.data.frame(colData(rld)[, "original_names"]), 1,
#      paste, collapse = ' : ')
# colnames(sampleDistMatrix) <- apply(as.data.frame(colData(rld)[, "original_names"]), 1,
#      paste, collapse = ' : ')

distances_df <- as.data.frame(colData(rld)[, intgroup, drop = F])
# apply(as.data.frame(colData(rld)[, params$design]), 1, paste, collapse = ' : ')

## Define colors to use for the heatmap
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

## Make the heatmap
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         color = colors,
         annotation_col = distances_df)

```

This plot shows how samples are clustered based on their euclidean distance using the regularized log transformed count data. This figure gives an overview of how the samples are hierarchically clustered. It is a complementary figure to the PCA plot.

## DEGs only (`r length(unique(significantResults[[biomart_filter]]))` genes)

```{r 'sampleDist_degs', fig.width = 10, fig.height = 10}
## Limit to the top n genes
# rv = rowVars(assay(rld))
# select = order(rv, decreasing = TRUE)[1:nBest]
## Obtain the sample euclidean distances
# sampleDistsTop <- dist(t(assay(rld)[select,]))
sampleDistsDEGs <- stats::dist(t(assay(rld_degs)))
sampleDistsDEGs <- dendextend::sort_dist_mat(sampleDistsDEGs)
# sampleDistMatrixTop <- as.matrix(sampleDistsTop)
sampleDistMatrixDEGs <- as.matrix(sampleDistsDEGs)

## Add names based on intgroup
# rownames(sampleDistMatrixDEGs) <- apply(as.data.frame(colData(rld)[, "original_names"]), 1,
#     paste, collapse = ' : ')
# colnames(sampleDistMatrixDEGs) <- apply(as.data.frame(colData(rld)[, "original_names"]), 1,
#     paste, collapse = ' : ')

distances_df <- as.data.frame(colData(rld)[, intgroup, drop = F])
# apply(as.data.frame(colData(rld)[, params$design]), 1, paste, collapse = ' : ')

## Define colors to use for the heatmap
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

## Make the heatmap
pheatmap(sampleDistMatrixDEGs,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         color = colors,
         annotation_col = distances_df)
```

This plot shows how samples are clustered based on their euclidean distance using the regularized log transformed count data of all DEGs. This figure gives an overview of how the samples are hierarchically clustered. It is a complementary figure to the PCA plot.

## Most `r nBest` variable genes only

```{r 'sampleDist_top', fig.width = 10, fig.height = 10}
## Limit to the top n genes
## Using the select object from PCA code above...
## Obtain the sample euclidean distances
sampleDistsTop <- stats::dist(t(assay(rld)[select,]))
sampleDistsTop <- dendextend::sort_dist_mat(sampleDistsTop)
sampleDistMatrixTop <- as.matrix(sampleDistsTop)

## Add names based on intgroup
# rownames(sampleDistMatrixTop) <- apply(as.data.frame(colData(rld)[, "original_names"]), 1,
#     paste, collapse = ' : ')
# colnames(sampleDistMatrixTop) <- apply(as.data.frame(colData(rld)[, "original_names"]), 1,
#     paste, collapse = ' : ')

distances_df <- as.data.frame(colData(rld)[, intgroup, drop = F])
# apply(as.data.frame(colData(rld)[, params$design]), 1, paste, collapse = ' : ')

## Define colors to use for the heatmap
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)

## Make the heatmap
pheatmap(sampleDistMatrixTop,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         color = colors,
         annotation_col = distances_df)
```

This plot shows how samples are clustered based on their euclidean distance using the regularized log transformed count data of the top `r nBest` most variable genes. This figure gives an overview of how the samples are hierarchically clustered. It is a complementary figure to the PCA plot.

# Heatmaps {.tabset .tabset-fade}

## All `r length(unique(significantResults[[biomart_filter]]))` DEGs

```{r 'heatmap_all_DEGs', collapse = TRUE, fig.height = 10}

mat <- assay(rld)[row.names(assay(rld)) %in%
                                          significantResults[[biomart_filter]],]
mat <- mat - rowMeans(mat)

heatmap_df <- as.data.frame(colData(rld)[, intgroup, drop = F])
row.names(heatmap_df) <- colData(rld)$original_names # Customize!
pheatmap(mat,
         annotation_col = heatmap_df,
         show_rownames = FALSE,
         border_color = NA,
         scale = "row")
# color = inferno(10)

```

## Top `r nHeatmapDEGs` differentially abundant genes

```{r 'heatmap_top_degs', fig.height = 10, fig.width = 10, collapse = TRUE}
mat_top <- assay(rld)[row.names(assay(rld)) %in% allResultsOrdered_logFC_filter[1:nHeatmapDEGs,][[biomart_filter]],]
mat_top <- mat_top - rowMeans(mat_top)

genes_for_heatmap <- data.frame(genes = row.names(mat_top))
names(genes_for_heatmap)[names(genes_for_heatmap) == "genes"] <- biomart_filter
genes_for_heatmap <- dplyr::left_join(genes_for_heatmap,
                                      id_table_entrez,
                                      by = biomart_filter) %>%
  dplyr::select(!!ensym(biomart_filter), !!ensym(species_gene_symbol)) %>%
  dplyr::distinct() %>%
  dplyr::group_by(!!ensym(biomart_filter)) %>%
  dplyr::mutate(combined = paste0(!!ensym(species_gene_symbol), collapse = ", ")) %>%
  dplyr::select(-!!ensym(species_gene_symbol)) %>%
  dplyr::distinct()
names(genes_for_heatmap)[names(genes_for_heatmap) == "combined"] <- species_gene_symbol

  
genes_for_heatmap[[species_gene_symbol]] <- ifelse(is.na(genes_for_heatmap[[species_gene_symbol]]),
               genes_for_heatmap[[biomart_filter]],
               genes_for_heatmap[[species_gene_symbol]])

genes_for_heatmap[[species_gene_symbol]] <- ifelse(genes_for_heatmap[[species_gene_symbol]] == "",
               genes_for_heatmap[[biomart_filter]],
               genes_for_heatmap[[species_gene_symbol]])

genes_for_heatmap <- genes_for_heatmap %>% pull(species_gene_symbol)

pheatmap(mat_top,
         annotation_col = heatmap_df,
         labels_row = genes_for_heatmap,
         show_rownames = TRUE,
         border_color = NA,
         scale = "row")
color = inferno(10)

```

## Top `r nHeatmap` variable genes

```{r 'heatmap_top_n_genes', fig.height = 10, fig.width = 10, collapse = TRUE}
rv = rowVars(assay(rld))
select = order(rv, decreasing = TRUE)[1:nHeatmap]
matRV <- assay(rld)[select,]
matRV <- matRV - rowMeans(matRV)

genes_for_heatmap <- data.frame(genes = row.names(matRV))
names(genes_for_heatmap)[names(genes_for_heatmap) == "genes"] <- biomart_filter
genes_for_heatmap <- dplyr::left_join(genes_for_heatmap,
                                      id_table_entrez,
                                      by = biomart_filter) %>%
  dplyr::select(!!ensym(biomart_filter), !!ensym(species_gene_symbol)) %>%
  dplyr::distinct() %>%
  dplyr::group_by(!!ensym(biomart_filter)) %>%
  dplyr::mutate(combined = paste0(!!ensym(species_gene_symbol), collapse = ", ")) %>%
  dplyr::select(-!!ensym(species_gene_symbol)) %>%
  dplyr::distinct()
names(genes_for_heatmap)[names(genes_for_heatmap) == "combined"] <- species_gene_symbol

  
genes_for_heatmap[[species_gene_symbol]] <- ifelse(is.na(genes_for_heatmap[[species_gene_symbol]]),
               genes_for_heatmap[[biomart_filter]],
               genes_for_heatmap[[species_gene_symbol]])

genes_for_heatmap[[species_gene_symbol]] <- ifelse(genes_for_heatmap[[species_gene_symbol]] == "",
               genes_for_heatmap[[biomart_filter]],
               genes_for_heatmap[[species_gene_symbol]])

genes_for_heatmap <- genes_for_heatmap %>% pull(species_gene_symbol)

pheatmap(matRV,
         #color = rev(brewer.pal(11, "RdBu")), # inferno(10),
         annotation_col = heatmap_df,
         labels_row = genes_for_heatmap,
         border_color = NA,
         scale = "row",
         cutree_rows = 3,
         cutree_cols = 4)

```


# MA plots

This section contains three groups of MA plots (see [Wikipedia](https://en.wikipedia.org/wiki/MA_plot)) that compare the mean of the normalized counts against the log fold change. Each of the groups has a tab for each contrast. The plots show one point per feature. The points are shown in red if the feature has an adjusted p-value less than the cutoff listed in each section, that is, the statistically significant features are shown in red.

## Filtered at `r metadata(resList[[1]])$alpha` {.tabset .tabset-fade}

This group of plots shows `alpha` = `r metadata(resList[[1]])$alpha`, which is the `alpha` value used to determine which resulting features were significant when running the function `DESeq2::results()`.  

```{r fc_range, warning = FALSE}
# Find IQR for absolute values of all significant log2foldchanges
# Plot MA plots
foldchangestats <- boxplot(abs(significantResults$log2FoldChange), plot = F)
ylims_for_ma_plots <- c(-foldchangestats$stats[4, 1], foldchangestats$stats[4, 1])
```

```{r 'MAplotalpha', results='asis'}
## MA plot with alpha used in DESeq2::results()

for (i in seq_along(resListAll)) {
  contrast = gsub(pattern = paste0("log2.*", DESIGN, "\ "),
                  replacement =  "",
                  x = resListAll[[i]]@elementMetadata[[2]][2])
  cat("###", contrast, "  \n\n")
  MA_title <- paste0('MA plot, alpha = ', metadata(resListAll[[i]])$alpha,', ', contrast)
  DESeq2::plotMA(resListAll[[i]],
                 alpha = metadata(resListAll[[i]])$alpha,
                 main = str_wrap(MA_title, width = 40),
                 ylim = ylims_for_ma_plots)
  cat('  \n\n')
}

## Used to have multiple tabsets for MA plots... it was too many MA plots.

## MA plot with alpha = 1/2 of the alpha used in DESeq2::results()
# for (i in 1:length(resList)) {
# contrast = gsub(pattern = "log2.*Time\ ",
#                 replacement =  "",
#                 x = resList[[i]]@elementMetadata[[2]][2])
# cat("###", contrast, " \n")
# DESeq2::plotMA(resList[[i]], alpha = metadata(resList[[i]])$alpha / 2,
#     main = paste('MA plot with alpha =', metadata(resList[[i]])$alpha / 2,',', contrast))
# cat('\n\n')
# }

## MA plot with alpha corresponding to the one that gives the nBest features
# for (i in 1:length(resList)) {
# nBest.actual <- min(nBest, nrow(head(resList[[i]], n = nBest)))
# nBest.alpha <- head( resList[[i]][order(resList[[i]]$pvalue),], n = nBest)$padj[nBest.actual]
# contrast = gsub(pattern = "log2.*Time\ ",
#                 replacement =  "",
#                 x = resList[[i]]@elementMetadata[[2]][2])
# cat("###", contrast, " \n")
# DESeq2::plotMA(resList[[i]], alpha = nBest.alpha * 1.00000000000001,
#     main = paste('MA plot for top', nBest.actual, 'features',',', contrast))
# cat('\n\n')
# }

```

## {-}

# P-values {.tabset .tabset-fade}

## Distribution of all p-values

```{r pvalueHistogram, collapse = TRUE}
## P-value histogram plot

ggplot(allResults, aes(x = pvalue)) +
    geom_histogram(alpha = .5, position = 'identity', bins = 50) +
    labs(title = 'Histogram of unadjusted p-values') +
    xlab('Unadjusted p-values') +
    facet_wrap( ~ contrast, ncol = 2)

```

This plot shows a histogram of the unadjusted p-values. It might be skewed right or left, or flat as shown in the [Wikipedia examples](https://en.wikipedia.org/wiki/Histogram#Examples). The shape depends on the percent of features that are differentially expressed. For further information on how to interpret a histogram of p-values check [David Robinson's post on this topic](http://varianceexplained.org/statistics/interpreting-pvalue-histogram/).

```{r pvalueSumm}
## P-value distribution summary
summary(allResults$pvalue)
#boxplot(allResults$pvalue)
```

This is the numerical summary of the distribution of the p-values.

```{r pvalueTable, results = 'asis'}
## Split features by different p-value cutoffs
pval_table <- lapply(c(1e-04, 0.001, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5,
    0.6, 0.7, 0.8, 0.9, 1), function(x) {
    data.frame('Cut' = x, 'Count' = sum(allResults$pvalue <= x, na.rm = TRUE))
})
pval_table <- do.call(rbind, pval_table)
kable(pval_table, format = 'markdown', align = c('c', 'c'))
```

This table shows the number of features with p-values less or equal than some commonly used cutoff values. 

## Distribution of adjusted p-values

```{r padjHistogram, collapse = TRUE}
## Adjusted p-values histogram plot
ggplot(allResults, aes(x = padj)) +
  geom_histogram(alpha = 0.5, position = 'identity', bins = 50) +
  labs(title = paste('Histogram of', elementMetadata(resList[[1]])$description[grep('adjusted', elementMetadata(resList[[1]])$description)])) +
  xlab('Adjusted p-values') +
  facet_wrap( ~ contrast, ncol = 2)
```

This plot shows a histogram of the `r elementMetadata(resList[[1]])$description[grep('adjusted', elementMetadata(resList[[1]])$description)]`. It might be skewed right or left, or flat as shown in the [Wikipedia examples](https://en.wikipedia.org/wiki/Histogram#Examples). 

```{r padjSumm}
## Adjusted p-values distribution summary
summary(res.df$padj)
```

This is the numerical summary of the distribution of the `r elementMetadata(resList[[1]])$description[grep('adjusted', elementMetadata(resList[[1]])$description)]`.

```{r padjTable, results = 'asis'}
## Split features by different adjusted p-value cutoffs
padj_table <- lapply(c(1e-04, 0.001, 0.01, 0.025, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5,
    0.6, 0.7, 0.8, 0.9, 1), function(x) {
    data.frame('Cut' = x, 'Count' = sum(res.df$padj <= x, na.rm = TRUE))
})
padj_table <- do.call(rbind, padj_table)
kable(padj_table, format = 'markdown', align = c('c', 'c'))
```

This table shows the number of features with `r elementMetadata(resList[[1]])$description[grep('adjusted', elementMetadata(resList[[1]])$description)]` less or equal than some commonly used cutoff values. 

# Tables of top features

This table shows the significant DEGs (passing all filtering criteria) ordered by their absolute fold change. Use the search function to find your feature of interest or sort by one of the columns. You can limit to a single contrast if desired.

```{r 'topFeatures', results = 'asis', warning = FALSE}
searchURL <- "http://www.ncbi.nlm.nih.gov/gene/?term="
## Add search url if appropriate
res.df.dt <- res.df
if (!is.null(searchURL)) {
    res.df.dt$ensembl_gene_id <- paste0('<a href="',
                             searchURL,
                             res.df.dt$ensembl_gene_id,
                             '" rel="noopener noreferrer" target="_blank">',
                             res.df.dt$ensembl_gene_id,
                             '<br/>',
                             res.df.dt[[species_gene_symbol]],
                             '</a>')
}

res.df.dt[, 'padj'] <- format(res.df.dt[, 'padj'], scientific = TRUE, digits = digits)
res.df.dt[, 'pvalue'] <- format(res.df.dt[, 'pvalue'], scientific = TRUE, digits = digits)
res.df.dt <- res.df.dt %>% dplyr::select(-c(description, !!ensym(species_gene_symbol)))

DT::datatable(res.df.dt,
          options = list(pagingType = 'full_numbers',
                         pageLength = 20,
                         scrollX = '100%',
                         dom = 'Bfrtip',
                         buttons = c('copy',
                                     'csv',
                                     'excel',
                                     'pdf',
                                     'print',
                                     'colvis'),
                         columnDefs = list(list(visible = FALSE, targets = c(2, 4, 5)))),
          escape = FALSE,
          extensions = 'Buttons',
          rownames = FALSE,
          filter = "top",
          colnames = c('Gene' = 'ensembl_gene_id')) %>% 
  DT::formatRound(which(!colnames(res.df.dt) %in% c('pvalue',
                                                    'padj',
                                                    'Feature',
                                                    'contrast',
                                                    'description',
                                                    'ensembl_gene_id',
                                                    'external_gene_name')),
                  digits)

```

# Gene-level plots for top `r nBestFeatures` features

This section contains plots showing the normalized counts per sample for each group of interest. Only the best `r nBestFeatures` features are shown, ranked by their absolute fold change values. The Y axis is on the log10 scale and the feature name is shown in the title of each plot.

```{r 'plotCounts'}
# Would be valuable to have this broken down by contrast too.
plotCounts_gg <- function(i, dds, intgroup) {
    plotdata <- plotCounts(dds,
                       gene = i,
                       intgroup = intgroup,
                       returnData = TRUE)
    plot_title <- paste(id_table[[biomart_filter]][id_table[[biomart_filter]] == i],
                        id_table[[species_gene_symbol]][id_table[[biomart_filter]] == i])
    if (ncol(plotdata) > 2) {
      colorCol = 3
    } else {colorCol = 2}
    ggplot(plotdata, aes(x = plotdata[, 2], y = plotdata[, 1], color = plotdata[, colorCol])) + 
      geom_point(position = position_jitterdodge()) +
      ylab('Normalized count') +
      xlab('Group') +
      ggtitle(plot_title) +
      coord_trans(y = "log10") +
      theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
      labs(color = colnames(plotdata)[colorCol])
}

genesToPlot <- significantResults %>% arrange(-abs(log2FoldChange))

for (i in head(unique(genesToPlot[[biomart_filter]]), nBestFeatures)) {
  print(plotCounts_gg(i, dds = dds, intgroup = params$design))
}

```

# Plots of genes of interest

This section shows genes of interest sorted by those with highest fold-change within each contrast.

```{r 'plot_genes_of_interest', fig.width = 10, fig.height = 6}

#numResults = 20
numResults <- nBestFeatures

allResultsOrdered_logFC_filter %>%
  group_by(contrast) %>%
  top_n(numResults, wt = abs(log2FoldChange)) %>%
  ungroup() %>%
  mutate(contrast = as.factor(contrast),
         symbol = reorder_within(!!ensym(species_gene_symbol), log2FoldChange, contrast)) %>%
  ggplot(aes(x = log2FoldChange,
             y = symbol,
             color = contrast,
             size = -log(padj))) +
  geom_point(show.legend = TRUE) +
  facet_wrap(~contrast,
             scales = "free_y",
             ncol = 4,
             labeller = labeller(contrast = label_wrap_gen(10))) +
  scale_y_reordered()  +
  geom_vline(xintercept = 0,
             linetype = "dashed",
             color = "black",
             size = 1) +
  ggtitle(paste0("Top ",
                 numResults,
                 " genes ranked by fold change (adjusted p-value <",
                 alpha,
                 "), grouped by treatment"))
```

# Volcano plot

This section shows a volcano plot for each contrast.  

Note that scales are set manually for this plot: therefore, there may be data points outside the range shown (see warnings).  

Data reported as significant are shown as red points.

```{r 'volcano-plot', fig.width = 12, fig.height = 6, warning = TRUE, collapse = TRUE}
ggplot(allResults, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(size = 0.5, alpha = 0.3) +
  geom_point(data = significantResults, size = 0.5, alpha = 0.9, color = "red") +
  facet_wrap(~contrast, ncol = 4) + # , scales = "free"
  geom_vline(xintercept = c(-log2(1.5), log2(1.5)), color = "red", alpha = 1.0) + 
  geom_hline(yintercept = -log10(0.05), color = "blue", alpha = 1.0) +
  scale_x_continuous(name = "log2 Fold Change", limits = c(-5, 5)) + # 
  scale_y_continuous(name = "-log10 adjusted p-value", limits = c(0, 6)) # 
```

```{r 'bmdexpress_input', collapse = TRUE}

###########################################################################################
######## PRODUCE INPUT FOR BMDExpress2 ####################################################
###########################################################################################
# Run this code conditional on whether there is a column matching the word dose in the metadata
if (any(grepl(x = colnames(DESeqDesign), pattern = "dose", ignore.case = T))) {
  doseCol <- grep(x = colnames(DESeqDesign), pattern = "dose", ignore.case = T)
  # Create input file...
  read.counts <- as.data.frame(counts(dds, normalized = TRUE))
  lognorm.read.counts <- log2(read.counts + 1)
  bmdexpress <- fastbmd <- as.data.frame(lognorm.read.counts)
  bmdexpress <- cbind(SampleID = c(row.names(bmdexpress)),
                      bmdexpress,
                      stringsAsFactors = F)
  bmdexpress <- rbind(c("Dose", as.character(DESeqDesign[colnames(bmdexpress)[-1],][, doseCol])),
                      bmdexpress,
                      stringsAsFactors = F)

  if (!is.na(params$group_facet)) {
    write.table(bmdexpress,
                file = file.path(paths$DEG_output,
                               paste0("bmdexpress_input_",
                                      paste(params$group_filter, collapse = "_"),
                                      ".txt")),
                quote = F,
                sep = "\t",
                row.names = F,
                col.names = T)
  } else {
    write.table(bmdexpress,
                file = file.path(paths$DEG_output, "bmdexpress_input.txt"),
                quote = F,
                sep = "\t",
                row.names = F,
                col.names = T)
  }
}
```

# GO Enrichment Analysis {.tabset .tabset-fade}

This section performs GO enrichment on **all DEGs passing filters**. The background set of genes is those that were identified in this sequencing experiment. The [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) package is used for this analysis.

```{r cluster-profiler, warning = FALSE, message = FALSE, collapse = TRUE}
## Remember that dds had ENSEMBL ids for the genes
# ensemblNames <- gsub("\\..*", "", rownames(dds))

ensemblDEGs <- significantResults[[biomart_filter]]
head(ensemblDEGs)
DEGs <- dplyr::left_join(data.frame(genes = ensemblDEGs), id_table_entrez, by = c("genes" = biomart_filter))
head(DEGs)
entrezDEGs <- as.character(DEGs$entrezgene_id)
entrezDEGs <- entrezDEGs[!is.na(entrezDEGs)]

#### DEFINE INPUTS
myDEGs <- entrezDEGs
# background <- row.names(assay(dds))
background <- dplyr::left_join(data.frame(genes = row.names(assay(dds))),
                               id_table_entrez,
                               by = c("genes" = biomart_filter)) %>%
  dplyr::filter(!is.na(entrezgene_id)) %>%
  dplyr::pull(entrezgene_id)
background <- as.character(background)

entrezDEGs <- entrezDEGs[!is.na(entrezDEGs)]

##### TO DO: Add multiple panels for each contrast.
##### Use "all" argument to only run each once..?

## Not all genes have a p-value
table(!is.na(resList[[1]]$padj))

KeyType <- "ENTREZID" # ENSEMBL? ACCNUM? GID? ENTREZID?
# head(keys(orgdb, keytype = "ENTREZID"))

# Run all at once...
# enrich_go_all <- enrichGO(gene = myDEGs,
#                          universe = background, # All genes in dataset
#                          OrgDb = orgdb,
#                          keyType = KeyType,
#                          readable = T,
#                          ont = "all",
#                          pAdjustMethod = "BH",
#                          pvalueCutoff = 0.01,
#                          qvalueCutoff = 0.05)

# clusterProfiler::dotplot(enrich_go_all, font.size = 9, showCategory = 10, split = "ONTOLOGY")  +
#   theme(axis.text.y = element_text(angle = 0)) +
#   scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 50)) +
#   facet_wrap(~ONTOLOGY)

## Can also do individually...
## Perform enrichment analysis for Biological Process (BP)
enrich_go_bp <- enrichGO(gene = myDEGs,
                         universe = background, # All genes in dataset
                         OrgDb = orgdb,
                         keyType = KeyType,
                         readable = T,
                         ont = "BP",
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.01,
                         qvalueCutoff = 0.05)

enrich_go_mf <- enrichGO(gene = myDEGs,
                         universe = background, # All genes in dataset
                         OrgDb = orgdb,
                         keyType = KeyType,
                         readable = T,
                         ont = "MF",
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.01,
                         qvalueCutoff = 0.05)

enrich_go_cc <- enrichGO(gene = myDEGs,
                         universe = background, # All genes in dataset
                         OrgDb = orgdb,
                         keyType = KeyType,
                         ont = "CC",
                         pAdjustMethod = "BH",
                         pvalueCutoff = 0.01,
                         qvalueCutoff = 0.05)

write.table(as.data.frame(enrich_go_bp),
            file = file.path(paths$pathway_analysis, "Enriched.GO_terms.Biological_Processes.txt"))
write.table(as.data.frame(enrich_go_mf),
            file = file.path(paths$pathway_analysis, "Enriched.GO_terms.Molecular_Functions.txt"))
write.table(as.data.frame(enrich_go_cc),
            file = file.path(paths$pathway_analysis, "Enriched.GO_terms.Cellular_Component.txt"))

```

## Biological processes

```{r calculate-foldchanges, message = FALSE, warning = FALSE}
  DEGs_full <- dplyr::left_join(significantResults,
                                id_table_entrez,
                                by = biomart_filter) %>%
    distinct()
  foldChanges <- DEGs_full %>% dplyr::pull(log2FoldChange)
  names(foldChanges) <- DEGs_full %>% dplyr::pull(entrezgene_id)
  foldChanges <- foldChanges %>% sort() %>% rev()
  foldChanges <- foldChanges[!is.na(names(foldChanges))]
```

```{r plot_GO_analysis_bp, message = FALSE, warning = FALSE, fig.width = 12, collapse = TRUE}
if (nrow(enrich_go_bp@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 3) {
  plot1 <- clusterProfiler::dotplot(enrich_go_bp, font.size = 9, showCategory = 20)  +
    theme(axis.text.y = element_text(angle = 0)) +
    scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))
  

  
  simplified_bp <- simplify(enrich_go_bp, cutoff = 0.7, by = "p.adjust", select_fun = min)
  # Max size of gene set... useful for visualizing networks
  simplified_bp_for_network <- simplified_bp %>% DOSE::gsfilter(by = "GSSize", max = 400)
  # I believe there is a bug in the current version of DOSE or clusterProfiler that prevents this from working
  # Throws this error on plotting: Error in graph.data.frame(x, directed = FALSE) : the data frame should contain at least two columns
  simplified_bp_filt <- simplified_bp %>%
      filter(p.adjust < .05, qvalue < 0.05) %>%
      mutate(GeneRatio = DOSE::parse_ratio(GeneRatio)) %>%
      arrange(desc(GeneRatio))
  
  show3 <- simplified_bp_filt$Description[1:3] # Top 3
  show3 <- show3[!is.na(show3)]
  show5 <- simplified_bp_filt$Description[1:5] # Top 5
  show5 <- show5[!is.na(show5)]
  show10 <- simplified_bp_filt$Description[1:10] # Top 10
  show10 <- show10[!is.na(show10)]
  
  plot2 <- cnetplot(simplified_bp,
           foldChange = foldChanges,
           showCategory = show10,
           cex_label_gene = 0.3)
  
  plot3 <- upsetplot(enrich_go_bp)
  plot4 <- heatplot(simplified_bp, foldChange = foldChanges, showCategory = show10) +
    scale_y_discrete(labels = function(x) stringr::str_trunc(x, width = 50))
  
  print(plot1)
  print(plot2)
  print(plot3)
  print(plot4)
} else {print("No significantly enriched terms using criteria selected") }
```

## Molecular Functions

```{r plot_GO_analysis_mf, message = FALSE, warning = FALSE, fig.width = 12, collapse = TRUE}
if (nrow(enrich_go_mf@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 3) {
  plot1 <- clusterProfiler::dotplot(enrich_go_mf, font.size = 9)  +
     theme(axis.text.y = element_text(angle = 0)) +
    scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))
  
  simplified_mf <- simplify(enrich_go_mf, cutoff = 0.7, by = "p.adjust", select_fun = min)
  # Max size of gene set... useful for visualizing networks
  simplified_mf_for_network <- simplified_mf %>% DOSE::gsfilter(by = "GSSize", max = 400)
  # I believe there is a bug in the current version of DOSE or clusterProfiler that prevents this from working
  # Throws this error on plotting: Error in graph.data.frame(x, directed = FALSE) : the data frame should contain at least two columns
  simplified_mf_filt <- simplified_mf %>%
      filter(p.adjust < .05, qvalue < 0.05) %>%
      mutate(GeneRatio = DOSE::parse_ratio(GeneRatio)) %>%
      arrange(desc(GeneRatio))
  
  show3 <- simplified_mf_filt$Description[1:3] # Top 3
  show5 <- simplified_mf_filt$Description[1:5] # Top 5
  show10 <- simplified_mf_filt$Description[1:10] # Top 10
  
  if (nrow(simplified_mf) > 3) {
  plot2 <- cnetplot(simplified_mf,
           foldChange = foldChanges,
           showCategory = show10,
           cex_label_gene = 0.3)
  }
  
  enrich_go_mf_upset <- enrich_go_mf
  enrich_go_mf_upset@result$Description <- str_trunc(enrich_go_mf_upset@result$Description, width = 50)
  plot3 <- upsetplot(enrich_go_mf_upset)
  plot4 <- enrichplot::heatplot(simplified_mf, foldChange = foldChanges, showCategory = show10) +
    scale_y_discrete(labels = function(x) stringr::str_trunc(x, width = 50))
  
  print(plot1)
  if (nrow(simplified_mf) > 3) {print(plot2)} else {print("Network plot not shown due to a limited number of enriched terms.")}
  print(plot3)
  print(plot4)
} else {print("No significantly enriched terms using criteria selected") }
```

## Cellular Component

```{r plot_GO_analysis_cc, message = FALSE, warning = FALSE, fig.width = 12, collapse = TRUE}
if (nrow(enrich_go_cc@result %>% filter(p.adjust < 0.01 & qvalue < 0.05)) > 3) {
  plot1 <- clusterProfiler::dotplot(enrich_go_cc, font.size = 9)  +
     theme(axis.text.y = element_text(angle = 0)) +
    scale_y_discrete(labels = function(x) stringr::str_wrap(x, width = 30))
  
  simplified_cc <- simplify(enrich_go_cc, cutoff = 0.7, by = "p.adjust", select_fun = min)
  # Max size of gene set... useful for visualizing networks
  simplified_cc_for_network <- simplified_cc %>% DOSE::gsfilter(by = "GSSize", max = 400)
  # I believe there is a bug in the current version of DOSE or clusterProfiler that prevents this from working
  # Throws this error on plotting: Error in graph.data.frame(x, directed = FALSE) : the data frame should contain at least two columns
  simplified_cc_filt <- simplified_cc %>%
      filter(p.adjust < .05, qvalue < 0.05) %>%
      mutate(GeneRatio = DOSE::parse_ratio(GeneRatio)) %>%
      arrange(desc(GeneRatio))
  
  show3 <- simplified_cc_filt$Description[1:3] # Top 3
  show5 <- simplified_cc_filt$Description[1:5] # Top 5
  show10 <- simplified_cc_filt$Description[1:10] # Top 10
  
  plot2 <- cnetplot(simplified_cc,
           foldChange = foldChanges,
           showCategory = show10,
           cex_label_gene = 0.3)
  
  enrich_go_cc_upset <- enrich_go_cc
  enrich_go_cc_upset@result$Description <- str_trunc(enrich_go_cc_upset@result$Description, width = 50)
  plot3 <- upsetplot(enrich_go_cc_upset)
  plot4 <- enrichplot::heatplot(simplified_cc, foldChange = foldChanges, showCategory = show10) +
    scale_y_discrete(labels = function(x) stringr::str_trunc(x, width = 50))
  
  print(plot1)
  print(plot2)
  print(plot3)
  print(plot4)
} else {print("No significantly enriched terms using criteria selected") }
```

```{r run_pathway_analysis_child_document, child = pathway_analysis, eval = params$run_pathway_analysis}
```

# Methods Summary

Please use this as a starting point for the bioinformatics/statistics section of any publications based on the data analyzed in this report. All of the information provided here is contained elsewhere in the report, but it is aggregated here for your convenience.  

There were `r samples_before` samples for which data was collected. A count matrix containing the $`r initialSampleDataCount`$ samples sequenced in this experiment was imported into R for statistical analysis. After removing samples with less than $`r threshold`$ reads, `r filteredSampleDataCount` samples were left. Following the exclusion of other samples *(which ones? Why? Please explain for your experiment, if applicable.)*, there were `r samples_after` samples remaining. Following the recommendations set out by the Omics Data Analysis Frameworks for Regulatory application (R-ODAF) guidelines, genes were filtered to include only those where 75% of at least one experimental group were above `r MinCount` CPM, and spurious spikes were removed in which (max - median) of counts were less than (sum of counts)/(number of replicates + 1). The samples excluded from analysis are shown in the table below. We used DESeq2 `r packageVersion("DESeq2")` [@ggplotcitation] to test for differentially abundant genes within the `r Platform` data. The log2FoldChange shrinkage procedure used was `r resList[[1]]@priorInfo$type`. An alpha of `r resList[[1]]@metadata$alpha` was used to extract raw results, which are reported as the `r gsub(mcols(x=resList[[1]])$description[[4]],pattern=":.*",replacement="")`. To account for multiple testing, `r mcols(resList[[1]])$description[5]` are reported. Cook's cutoff was set to `r cooks` in this analysis. Differentially expressed genes (DEGs) were filtered using a linear fold change cutoff of `r linear_fc_filter` and adjusted p-value of `r alpha`.  

## References {-}

If these methods were used in your study, please cite the following papers as appropriate:  

---
nocite: '@*'
---

<div id="refs"></div>

# Session Info

## Date the report was generated.

```{r reproducibility1, echo = FALSE}
## Date the report was generated
Sys.time()
```

## Parameters Used

From the list elements in each params${variable} used to generate this report.

```{r paramsList, echo = FALSE}
df <- as.data.frame(unlist(params))
names(df) <- "Parameter Value"
knitr::kable(as.data.frame(df), format = "markdown")
```

## Wallclock time spent generating the report.

```{r reproducibility2, echo = FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

## `R` session information.

```{r reproducibility3, echo = FALSE}
## Session info
options(width = 120)
session_info()
```

## Pandoc version used: `r rmarkdown::pandoc_version()`.

```{r 'save_final', show = FALSE}
# Save complete workspace
if (is.na(params$group_facet)) {
  save.image(file = file.path(paths$RData, "Complete_analysis.RData"))
} else {
  save.image(file = file.path(paths$RData,
                              paste0("Complete_analysis_",
                                     paste(params$group_filter, collapse = "_"),
                                     ".RData")))
}
# Add a flag to the start of the script with if statements in all code chunks to check if the flag is set.
# If flag is TRUE, run analysis; if flag is FALSE, SKIP analysis and continue here (just to modify plotting output).
```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
